witness/golang/omniwitness/usbarmory/internal/storage/slots/journal.go

```diff
@@ -0,0 +1,295 @@
+	}
+	e := &entry{
+		Magic:      [4]byte{magic0[0], magic0[1], magic0[2], magic0[3]},
+		Revision:   j.current.Revision + 1,
+		DataLen:    uint64(len(data)),
+		DataSHA256: sha256.Sum256(data),
+		Data:       data,
+	}
+
+	cap := (j.start + j.length - j.nextBlock) * j.dev.BlockSize()
+	if cap < uint(e.Size()) {
+		// The record won't fit in the remaining space, so wrap around and write at the beginning.
+		j.nextBlock = j.start
+	}
+	// TODO(al): consider making this more "streamy".
+	buf := &bytes.Buffer{}
+	if err := marshalEntry(e, buf); err != nil {
+		return fmt.Errorf("failed to marshal entry: %v", err)
+	}
+	return j.dev.WriteBlocks(j.nextBlock, buf.Bytes())
+}
+
+// Init scans the journal to figure out the latest valid record, if any.
+func (j *Journal) init() error {
+	// Start where all good stories do: at the beginning!
+	lba := j.start
+	var lastEntry entry
+	nextWriteLBA := j.start
+	for lba < j.start+j.length {
+		br := newBlockReader(j.dev, lba)
+		e, err := unmarshalEntry(br)
+		if err != nil {
+			if lastEntry.Revision > 0 {
+				// We already found the lastet record in the journal, so we're done.
+				break
+			}
+			// Since we haven't already found a good entry, and we were unable
+			// to unmarshal this one,
+			// this means that either:
+			//  a) the journal is completely empty, or
+			//  b) the previously good entry/ies at the start of the journal
+			//     have been completely or partially overwritten during a
+			//     failed write attempt.
+			// Either way, we don't have a valid entry wth a length field we can
+			// rely on, so we we'll have to fall back to scanning all blocks to
+			// look for one.
+			lba++
+			continue
+		}
+		if lastEntry.Revision == 0 || e.Revision > lastEntry.Revision {
+			// We've found a(nother) good entry, so update our state
+			lastEntry = *e
+			// Skip past the blocks we've just read
+			lba = (br.pos-1)/br.dev.BlockSize() + 1
+			// If this turns out to be the last good entry, then we'll write
+			// at the next block.
+			nextWriteLBA = lba
+			// But loop around again, just in case there are yet more good
+			// entries following on...
+			continue
+		} else if e.Revision < lastEntry.Revision {
+			// We've found an older revision following a newer one, so we're done.
+			nextWriteLBA = lba
+			break
+		} else {
+			return fmt.Errorf("journal is corrupt - found two entries with the same revision (%d)", e.Revision)
+		}
+	}
+	// In the case where the very last entry in the journal is the current one,
+	// and that entry extends into the final block, we'll need to wrap the
+	// nextBlock pointer to the start of the journal.
+	if nextWriteLBA >= j.start+j.length {
+		nextWriteLBA = j.start
+	}
+	j.nextBlock = nextWriteLBA
+	j.current = lastEntry
+
+	return nil
+}
+
+// unmarshalEntry reads and deserialises an entry structure from the provided reader.
+func unmarshalEntry(r io.Reader) (*entry, error) {
```

Comment:
Is there a reason to return a pointer here? Similarly with accepting it below.

I thought the go preference was generally to copy structs unless there's a good reason not to. But you've been in this game longer than me.

Reply:
The main reason is just to not return a entry in some unknown/invalid state in the case of error unless that's a specific requirement (which would warrant a comment on the func about this behaviour) - this is mostly the pattern followed by the Go std lib (a notable exception is probably the `Reader` types which can return valid data and error, but that's because it can return an EOF "error").